src/Database/Driver/Elasticsearch.php

```diff
@@ -0,0 +1,336 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @internal      Should be used internal only - will be replaced with a real driver in
+ * the future.
+ * @since         2.1.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\ElasticSearch\Database\Driver;
+
+use Cake\Database\Driver;
+use Cake\Database\Query;
+use Cake\Database\ValueBinder;
+use Elastica\Client as ElasticaClient;
+
+/**
+ * Represents a elasticsearch driver containing all specificities for
+ * a elasticsearch engine including its DSL dialect.
+ */
+class Elasticsearch extends Driver
+{
+    /**
+     * Base configuration settings for ElasticSearch driver
+     *
+     * @var array
+     */
+    protected $_baseConfig = [
+        'host' => 'localhost',
+        'username' => null,
+        'password' => null,
+        'port' => 9200,
+    ];
+
+    /**
+     * Elastica client instance
+     *
+     * @var \Elastica\Client;
+     */
+    protected $_client;
+
+    /**
+     * Constructor.
+     *
+     * @param array $config config options
+     * @param callable $callback Callback function which can be used to be notified
+     * about errors (for example connection down)
+     */
+    public function __construct(array $config = [], $callback = null)
+    {
+        $config += $this->_baseConfig;
+        $this->_config = $config;
+
+        $esLogger = $config['esLogger'];
```

Comment:
Any reason to not use `logger`?

Reply:
As current `Connection` have both `logger()` and `getEsLogger()` (and both properties), I tried maintain it.

I really don't know if we need both loggers to be configurable, as the `esLogger` is just a wrapper around `logger`.

Reply:
Sorry, I had forgotten about the two loggers. We can clean that up separately.